Project Walking/Source code/app/src/main/java/com/teamapricot/projectwalking/LocationHandler.java

```diff
@@ -0,0 +1,153 @@
+        locationManager = (LocationManager) activity.getApplicationContext()
+                .getSystemService(AppCompatActivity.LOCATION_SERVICE);
+        updateListeners = new ArrayList<>();
+        updateInterval = interval;
+
+        if (permissionHandler.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
+            initializeLocationHandler();
+        } else {
+            permissionHandler.requestPermissionAsync(Manifest.permission.ACCESS_FINE_LOCATION)
+                    .thenAccept(permissionGranted -> {
+                        if (permissionGranted) {
+                            initializeLocationHandler();
+                        } else {
+                            // TODO: Permission not granted, do something!
+                            System.out.println("permission not granted");
+                        }
+                    });
+        }
+    }
+
+    public Double getLatitude() {
+        if (position == null) {
+            return null;
+        }
+        return position.getLatitude();
+    }
+
+    public Double getLongitude() {
+        if (position == null) {
+            return null;
+        }
+        return position.getLongitude();
+    }
+
+    public Double getAccuracy() {
+        if (position == null) {
+            return null;
+        }
+        return position.getAccuracy();
+    }
+
+    /**
+     * Register an UpdateListener to get location updates.
+     *
+     * @param listener The listener to register
+     */
+    public void registerUpdateListener(UpdateListener listener) {
+        updateListeners.add(listener);
+    }
+
+    /**
+     * Unregister an UpdateListener.
+     *
+     * @param listener The listener to unregister
+     */
+    public void unregisterUpdateListener(UpdateListener listener) {
+        updateListeners.remove(listener);
+    }
+
+    private LocationListener createLocationListener() {
+        return location -> {
+            double latitude = location.getLatitude();
+            double longitude = location.getLongitude();
+            double accuracy = location.getAccuracy();
+
+            position = new Position(latitude, longitude, accuracy);
+
+            for (UpdateListener l : updateListeners) {
+                l.onPositionUpdated();
+            }
+        };
+
+    }
+
+    private void initializeLocationHandler() {
+        if (locationListener == null) {
+            locationListener = createLocationListener();
+        }
+
+        if (permissionHandler.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
+            locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER,
+                    updateInterval, 0, locationListener);
+        }
+    }
+
+    /**
+     * The interface to use for receiving location updates.
+     */
+    public interface UpdateListener {
+        void onPositionUpdated();
+    }
+
+    /**
+     * A position.
+     */
+    public static class Position {
```

Comment:
Right now this class is not used outside of the `LocationHandler` class. Is there a reason why it is `public`?

Reply:
The reason of me being sloppy. It should be private.

Reply:
And back to public.